File: android-compose/app/src/main/java/com/ohmz/tday/compose/core/calendar/CalendarSyncManager.kt
package com.ohmz.tday.compose.core.calendar

import com.ohmz.tday.compose.core.data.cache.OfflineCacheManager
import com.ohmz.tday.compose.core.data.todo.TodoRepository
import com.ohmz.tday.compose.core.model.TodoItem
import com.ohmz.tday.compose.core.model.TodoListMode
import com.ohmz.tday.compose.core.observability.TdayTelemetry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.TimeZone
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Mirrors pending scheduled tasks into T'Day's own device calendar.
 *
 * One-way by design: T'Day is the source of truth and the calendar is a projection of it. Nothing
 * is ever read back, so there is no conflict resolution and no chance of an edit loop.
 *
 * Reconciles the whole calendar rather than tracking per-task mutations, mirroring
 * `TaskReminderScheduler.rescheduleAll()`. That keeps the trigger surface to a single observation
 * of [OfflineCacheManager.cacheDataVersion] instead of the eight scattered call sites the reminder
 * scheduler needs, and it self-heals when the calendar drifts.
 *
 * Floaters are excluded: they have no due date, and the calendar contract keeps them off calendar
 * surfaces.
 */
@Singleton
class CalendarSyncManager @Inject constructor(
    private val repository: TodoRepository,
    private val cacheManager: OfflineCacheManager,
    private val deviceCalendarStore: DeviceCalendarStore,
    private val preferenceStore: CalendarSyncPreferenceStore,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val syncMutex = Mutex()
    private val observing = AtomicBoolean(false)

    /**
     * Starts mirroring on every cache-data change. Safe to call more than once; only the first
     * call attaches the observer.
     */
    fun start() {
        if (!observing.compareAndSet(false, true)) return
        scope.launch {
            // drop(1): the StateFlow replays its current value on collect, and the explicit
            // startup sync below already covers that first emission.
            cacheManager.cacheDataVersion.drop(1).collect {
                runCatching { syncNow() }
            }
        }
        scope.launch { runCatching { syncNow() } }
    }

    /**
     * Turns the mirror on and performs the first sync. The caller owns the permission grant.
     */
    suspend fun enable() {
        preferenceStore.setEnabled(true)
        // Force a write even when the task set matches the stored fingerprint: the calendar may
        // have been deleted while the feature was off.
        preferenceStore.setLastSyncedFingerprint(null)
        syncNow()
    }

    /** Turns the mirror off and removes the calendar and every event in it. */
    suspend fun disable() {
        preferenceStore.setEnabled(false)
        preferenceStore.setLastSyncedFingerprint(null)
        syncMutex.withLock {
            runCatching { deviceCalendarStore.deleteCalendar() }
                .onFailure { TdayTelemetry.capture(it, "calendar.sync.delete_calendar") }
        }
        TdayTelemetry.addBreadcrumb("calendar.sync.disabled")
    }

    /**
     * Reconciles the device calendar against the current task cache. A no-op when the feature is
     * off, when calendar permission is missing, or when nothing this mirror renders has changed.
     */
    suspend fun syncNow() {
        if (!preferenceStore.isEnabled()) return
        if (!deviceCalendarStore.hasPermissions()) return

        syncMutex.withLock {
            val events = buildEvents()
            val fingerprint = fingerprintOf(events)
            if (fingerprint == preferenceStore.getLastSyncedFingerprint()) return

            val calendarId = deviceCalendarStore.ensureCalendarId()
            if (calendarId == null) {
                TdayTelemetry.addBreadcrumb("calendar.sync.calendar_unavailable")
                return
            }

            runCatching { deviceCalendarStore.replaceEvents(calendarId, events) }
                .onSuccess {
                    preferenceStore.setLastSyncedFingerprint(fingerprint)
                    TdayTelemetry.addBreadcrumb(
                        "calendar.sync.reconcile",
                        data = mapOf("eventCount" to events.size),
                    )
                }
                .onFailure {
                    // Leave the fingerprint untouched so the next pass retries.
                    TdayTelemetry.capture(it, "calendar.sync.reconcile")
                }
        }
    }

    private fun buildEvents(): List<DeviceCalendarEvent> {
        val timeZoneId = TimeZone.getDefault().id
        return runCatching { repository.fetchTodosSnapshot(TodoListMode.ALL) }
            .getOrElse { emptyList() }
            .filterNot { it.completed }
            .mapNotNull { task -> task.toDeviceEvent(timeZoneId) }
    }

    private fun TodoItem.toDeviceEvent(timeZoneId: String): DeviceCalendarEvent? {
        val startMillis = due?.toEpochMilli() ?: return null
        if (startMillis <= 0) return null
        return DeviceCalendarEvent(
            taskId = id,
            title = title,
            description = description,
            startEpochMillis = startMillis,
            rrule = rrule?.takeIf { it.isNotBlank() },
            timeZoneId = timeZoneId,
        )
    }

    /**
     * Content fingerprint of the projected calendar. Cheap guard against rewriting the calendar
     * for cache changes that touched nothing this mirror renders — a floater edit, a completed-item
     * refresh, a preference sync.
     */
    private fun fingerprintOf(events: List<DeviceCalendarEvent>): String {
        if (events.isEmpty()) return "empty"
        return events
            .sortedWith(compareBy({ it.startEpochMillis }, { it.taskId }))
            .joinToString("|") { event ->
                listOf(
                    event.taskId,
                    event.startEpochMillis.toString(),
                    event.title,
                    event.description.orEmpty(),
                    event.rrule.orEmpty(),
                    event.timeZoneId,
                ).joinToString(RECORD_SEPARATOR)
            }
            .hashCode()
            .toString()
    }

    private companion object {
        /** Field separator that cannot appear in a task title, so fields cannot alias. */
        const val RECORD_SEPARATOR = "\u001F"
    }
}

File: ios-swiftUI/Tday/Core/Calendar/CalendarSyncManager.swift
import Foundation

/// Mirrors pending scheduled tasks into T'Day's own device calendar.
///
/// One-way by design: T'Day is the source of truth and the calendar is a projection of it. Nothing
/// is ever read back into the app, so there is no conflict resolution and no chance of an edit loop.
///
/// Reconciles the whole calendar rather than tracking per-task mutations, mirroring
/// `TaskReminderScheduler.reschedule(tasks:defaultReminder:)` — which is also why this hangs off the
/// same `AppViewModel.rescheduleReminders()` pass instead of growing its own trigger points.
///
/// Floaters are excluded: they have no due date, and the calendar contract keeps them off calendar
/// surfaces.
@MainActor
final class CalendarSyncManager {
    private let deviceCalendarStore: DeviceCalendarStore
    private let preferenceStore: CalendarSyncPreferenceStore

    init(deviceCalendarStore: DeviceCalendarStore, preferenceStore: CalendarSyncPreferenceStore) {
        self.deviceCalendarStore = deviceCalendarStore
        self.preferenceStore = preferenceStore
    }

    var isEnabled: Bool {
        preferenceStore.isEnabled
    }

    /// Prompts for calendar access and, if granted, turns the mirror on. Returns whether it is on.
    func enable(tasks: [TodoItem]) async -> Bool {
        guard await deviceCalendarStore.requestAccess() else {
            TdayTelemetry.addBreadcrumb("calendar.sync.permission_denied")
            return false
        }
        preferenceStore.setEnabled(true)
        // Force a write even when the task set matches the stored fingerprint: the calendar may
        // have been deleted while the feature was off.
        preferenceStore.setLastSyncedFingerprint(nil)
        await sync(tasks: tasks)
        return true
    }

    /// Turns the mirror off and removes the calendar and every event in it.
    func disable() async {
        preferenceStore.setEnabled(false)
        do {
            try deviceCalendarStore.deleteCalendar()
        } catch {
            TdayTelemetry.capture(error, operation: "calendar.sync.delete_calendar")
        }
        TdayTelemetry.addBreadcrumb("calendar.sync.disabled")
    }

    /// Reconciles the device calendar against `tasks`. A no-op when the feature is off, when
    /// calendar access is missing, or when nothing this mirror renders has changed.
    func sync(tasks: [TodoItem]) async {
        guard preferenceStore.isEnabled, deviceCalendarStore.hasAccess else { return }

        let events = Self.events(from: tasks)
        let fingerprint = Self.fingerprint(of: events)
        guard fingerprint != preferenceStore.lastSyncedFingerprint else { return }

        do {
            let calendar = try deviceCalendarStore.ensureCalendar()
            try deviceCalendarStore.replaceEvents(events, in: calendar)
            preferenceStore.setLastSyncedFingerprint(fingerprint)
            TdayTelemetry.addBreadcrumb(
                "calendar.sync.reconcile",
                data: ["eventCount": events.count]
            )
        } catch {
            // Leave the fingerprint untouched so the next pass retries.
            TdayTelemetry.capture(error, operation: "calendar.sync.reconcile")
        }
    }

    // MARK: - Projection

    static func events(from tasks: [TodoItem]) -> [DeviceCalendarEvent] {
        tasks.compactMap { task in
            guard !task.completed, let due = task.due else { return nil }
            let rrule = task.rrule.flatMap { $0.isEmpty ? nil : $0 }
            return DeviceCalendarEvent(
                taskID: task.id,
                title: task.title,
                notes: task.description,
                start: due,
                rrule: rrule
            )
        }
    }

    /// Content fingerprint of the projected calendar. Cheap guard against rewriting the calendar
    /// for refreshes that touched nothing this mirror renders — a floater edit, a completed-item
    /// refresh, a preference sync.
    static func fingerprint(of events: [DeviceCalendarEvent]) -> String {
        if events.isEmpty { return "empty" }
        // Field separator that cannot appear in a task title, so fields cannot alias.
        let separator = "\u{1F}"
        return events
            .sorted { lhs, rhs in
                lhs.start == rhs.start ? lhs.taskID < rhs.taskID : lhs.start < rhs.start
            }
            .map { event in
                [
                    event.taskID,
                    String(Int(event.start.timeIntervalSince1970)),
                    event.title,
                    event.notes ?? "",
                    event.rrule ?? ""
                ].joined(separator: separator)
            }
            .joined(separator: "|")
            .stableHash()
    }
}

private extension String {
    /// FNV-1a. Swift's own `hashValue` is seeded per process launch, so it would produce a
    /// different fingerprint on every cold start and force a pointless calendar rewrite each time
    /// the app is relaunched. This has to stay stable across launches to be worth anything.
    func stableHash() -> String {
        var hash: UInt64 = 0xcbf2_9ce4_8422_2325
        for byte in utf8 {
            hash ^= UInt64(byte)
            hash = hash &* 0x0000_0100_0000_01b3
        }
        return String(hash, radix: 16)
    }
}
